File: docs/examples/javascript/dae_collocation.js
//
//     MIT No Attribution
//
//     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
//
//     Permission is hereby granted, free of charge, to any person obtaining a copy of this
//     software and associated documentation files (the "Software"), to deal in the Software
//     without restriction, including without limitation the rights to use, copy, modify,
//     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
//     permit persons to whom the Software is furnished to do so.
//
//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//
// JS port of docs/examples/python/dae_collocation.py.
//
// Hand-built direct collocation of a crane/pendulum DAE optimal control
// problem (Mario Zanon & Sebastien Gross, KU Leuven 2012), solved as one
// big NLP with ipopt.  No integrator plugin is used -- the collocation
// equations are assembled by hand -- so this ports directly.  The original
// ipopt option linear_solver='ma27' is dropped (default mumps is used) and
// ipopt.max_iter is left at its default (see the note near the solve).
//
// JS notes (see README.md):
//   * numpy linear algebra (inv, @) is replaced by casadi DM ops (ca.inv,
//     ca.mtimes).  Collocation coefficients C/D are plain JS arrays.
//   * The NLP vector V is sliced via per-scalar vertsplit + range vertcat.
//   * Plotting is dropped; the optimal cost and a few state samples logged.

async function example(ca, log) {
  const inf = Infinity;

  // -------- Collocation setup --------
  const nicp = 1;
  const xref = 0.1;
  const l = 1.0, mmass = 1.0, Mmass = 1.0, g = 9.81;
  const tf = 5.0;
  const nk = 50;
  const ndstate = 6, nastate = 1, ninput = 1;
  const deg = 4;
  const h = tf / nk / nicp;

  const tau = ca.SX.sym("tau");
  const tau_root = [0].concat(ca.collocation_points(deg, "radau"));

  // Lagrange-polynomial collocation coefficients
  const C = Array.from({ length: deg + 1 }, () => new Array(deg + 1).fill(0));
  const D = new Array(deg + 1).fill(0);
  for (let j = 0; j <= deg; ++j) {
    let L = ca.SX(1);
    for (let j2 = 0; j2 <= deg; ++j2) {
      if (j2 !== j) L = ca.times(L, ca.rdivide(ca.minus(tau, tau_root[j2]), tau_root[j] - tau_root[j2]));
    }
    const lfcn = ca.Function("lfcn", [tau], [L]);
    D[j] = Number(lfcn(1.0));
    const tfcn = ca.Function("tfcn", [tau], [ca.tangent(L, tau)]);
    for (let j2 = 0; j2 <= deg; ++j2) C[j][j2] = Number(tfcn(tau_root[j2]));
  }

  // -------- Model setup (implicit DAE) --------
  const t = ca.SX.sym("t");
  const u = ca.SX.sym("u");
  const xd = ca.SX.sym("xd", ndstate);
  const xa = ca.SX.sym("xa", nastate);
  const xddot = ca.SX.sym("xdot", ndstate);
  const p = ca.SX.sym("p", 0, 1);

  const [x, y, w, dx, dy, dw] = ca.vertsplit(xd);
  const [xa0] = ca.vertsplit(xa);
  const xdd = ca.vertsplit(xddot);

  const res = ca.vertcat(
    ca.minus(xdd[0], dx),
    ca.minus(xdd[1], dy),
    ca.minus(xdd[2], dw),
    ca.plus(ca.times(mmass, xdd[3]), ca.times(ca.minus(x, w), xa0)),
    ca.minus(ca.plus(ca.times(mmass, xdd[4]), ca.times(y, xa0)), g * mmass),
    ca.plus(ca.plus(ca.times(Mmass, xdd[5]), ca.times(ca.minus(w, x), xa0)), u),
    ca.plus(ca.plus(ca.plus(
      ca.times(ca.minus(x, w), ca.minus(xdd[3], xdd[5])),
      ca.times(y, xdd[4])),
      ca.times(dy, dy)),
      ca.times(ca.minus(dx, dw), ca.minus(dx, dw))));

  const ffcn = ca.Function("ffcn", [t, xddot, xd, xa, u, p], [res]);
  const MayerTerm = ca.Function("mayer", [t, xd, xa, u, p],
    [ca.plus(ca.plus(ca.plus(
      ca.times(ca.minus(x, xref), ca.minus(x, xref)),
      ca.times(ca.minus(w, xref), ca.minus(w, xref))),
      ca.times(dx, dx)), ca.times(dy, dy))]);
  const LagrangeTerm = ca.Function("lagrange", [t, xd, xa, u, p],
    [ca.plus(
      ca.times(ca.minus(x, xref), ca.minus(x, xref)),
      ca.times(ca.minus(w, xref), ca.minus(w, xref)))]);

  // Bounds
  const u_min = [-2], u_max = [2];
  const xD_min = [-inf, -inf, -inf, -inf, -inf, -inf];
  const xD_max = [inf, inf, inf, inf, inf, inf];
  const xDi_min = [0.0, l, 0.0, 0.0, 0.0, 0.0];
  const xDi_max = [0.0, l, 0.0, 0.0, 0.0, 0.0];
  const xD_init = [0.0, l, 0.0, 0.0, 0.0, 0.0];
  const xA_min = [-inf], xA_max = [inf];
  const xA_init = [Math.sign(l) * 9.81];

  // -------- NLP variable layout --------
  const nx = ndstate + nastate;
  const ndiff = ndstate, nalg = nastate, nu = ninput, NP = 0;
  const NXD = nicp * nk * (deg + 1) * ndiff;
  const NXA = nicp * nk * deg * nalg;
  const NU = nk * nu;
  const NXF = ndiff;
  const NV = NXD + NXA + NU + NXF + NP;

  const V = ca.MX.sym("V", NV);
  const Vrows = ca.vertsplit(V, Array.from({ length: NV + 1 }, (_, i) => i));
  const slice = (a, b) => ca.vcat(Vrows.slice(a, b));

  const vars_lb = new Array(NV).fill(0);
  const vars_ub = new Array(NV).fill(0);
  const vars_init = new Array(NV).fill(0);
  const setRange = (arr, off, vals) => { for (let i = 0; i < vals.length; ++i) arr[off + i] = vals[i]; };

  // XD[k][i][j], XA[k][i][j-1], U[k]
  const XD = Array.from({ length: nk + 1 }, () => Array.from({ length: nicp }, () => new Array(deg + 1).fill(null)));
  const XA = Array.from({ length: nk }, () => Array.from({ length: nicp }, () => new Array(deg).fill(null)));
  const U = new Array(nk).fill(null);

  let offset = 0;
  for (let k = 0; k < nk; ++k) {
    for (let i = 0; i < nicp; ++i) {
      for (let j = 0; j <= deg; ++j) {
        XD[k][i][j] = slice(offset, offset + ndiff);
        if (j !== 0) XA[k][i][j - 1] = slice(offset + ndiff, offset + ndiff + nalg);
        if (k === 0 && j === 0 && i === 0) {
          setRange(vars_init, offset, xD_init);
          setRange(vars_lb, offset, xDi_min);
          setRange(vars_ub, offset, xDi_max);
          offset += ndiff;
        } else if (j !== 0) {
          setRange(vars_init, offset, xD_init.concat(xA_init));
          setRange(vars_lb, offset, xD_min.concat(xA_min));
          setRange(vars_ub, offset, xD_max.concat(xA_max));
          offset += nx;
        } else {
          setRange(vars_init, offset, xD_init);
          setRange(vars_lb, offset, xD_min);
          setRange(vars_ub, offset, xD_max);
          offset += ndiff;
        }
      }
    }
    U[k] = slice(offset, offset + nu);
    setRange(vars_lb, offset, u_min);
    setRange(vars_ub, offset, u_max);
    setRange(vars_init, offset, [0.0]);
    offset += nu;
  }
  XD[nk][0][0] = slice(offset, offset + ndiff);
  setRange(vars_lb, offset, xD_min);
  setRange(vars_ub, offset, xD_max);
  setRange(vars_init, offset, xD_init);
  offset += ndiff;
  if (offset !== NV) throw new Error("offset != NV (" + offset + " vs " + NV + ")");

  const P = ca.MX(0, 1);

  // -------- Constraints --------
  const gcon = [];
  // Collocation + continuity equations
  for (let k = 0; k < nk; ++k) {
    for (let i = 0; i < nicp; ++i) {
      for (let j = 1; j <= deg; ++j) {
        let xp_jk = ca.MX.zeros(ndiff);
        for (let j2 = 0; j2 <= deg; ++j2) xp_jk = ca.plus(xp_jk, ca.times(C[j2][j], XD[k][i][j2]));
        const fk = ffcn(ca.DM(0), ca.rdivide(xp_jk, h), XD[k][i][j], XA[k][i][j - 1], U[k], P);
        const fkrows = ca.vertsplit(fk, [0, ndiff, ndiff + nalg]);
        gcon.push(fkrows[0]); // differential part == 0
        gcon.push(fkrows[1]); // algebraic part == 0
      }
      let xf_k = ca.MX.zeros(ndiff);
      for (let j = 0; j <= deg; ++j) xf_k = ca.plus(xf_k, ca.times(D[j], XD[k][i][j]));
      if (i === nicp - 1) gcon.push(ca.minus(XD[k + 1][0][0], xf_k));
      else gcon.push(ca.minus(XD[k][i + 1][0], xf_k));
    }
  }

  // -------- Objective --------
  // Mayer term at the final collocation node
  let Obj = MayerTerm(ca.DM(0), XD[nk - 1][0][deg], XA[nk - 1][0][deg - 1], U[nk - 1], P);

  // Lagrange term via the collocation quadrature weights
  // lDotAtTauRoot = C.T ; ldInv = inv(C.T[1:,1:]) ; lAtOne = D
  const CT = Array.from({ length: deg + 1 }, (_, a) => Array.from({ length: deg + 1 }, (_, b) => C[b][a]));
  const ldInv = ca.MX(ca.inv(ca.DM(CT.slice(1).map((row) => row.slice(1)))));
  const lAtOne1 = ca.MX(ca.DM(D.slice(1)));

  for (let k = 0; k < nk; ++k) {
    for (let i = 0; i < nicp; ++i) {
      const dqParts = [];
      for (let j = 1; j <= deg; ++j) {
        dqParts.push(LagrangeTerm(ca.DM(0), XD[k][i][j], XA[k][i][j - 1], U[k], P));
      }
      const dQs = ca.times(h, ca.vcat(dqParts)); // deg x 1
      const Qs = ca.mtimes(ldInv, dQs);                       // deg x 1
      Obj = ca.plus(Obj, ca.mtimes(ca.transpose(Qs), lAtOne1)); // scalar
    }
  }
  // -------- Solve --------
  // NB: integer-valued ipopt sub-options (e.g. ipopt.max_iter) are mis-cast
  // by the wasm option bridge and clamp the solve to 1 iteration, so we rely
  // on the default max_iter here.
  const nlp = { x: V, f: Obj, g: ca.vcat(gcon) };
  const solver = ca.nlpsol("solver", "ipopt", nlp,
    { expand: true, "ipopt.tol": 1e-4, "ipopt.print_level": 0, print_time: false });

  const sol = solver.call({
    x0: ca.DM(vars_init), lbx: ca.DM(vars_lb), ubx: ca.DM(vars_ub), lbg: ca.DM(0), ubg: ca.DM(0),
  });

  log("optimal cost: " + sol["f"]);

  // Chariot position x (xd[0]) at the start of each finite element: the
  // per-element layout has a fixed stride, so a strided slice does it.
  const stride = nicp * ((deg + 1) * ndiff + deg * nalg) + nu;
  const xstarts = sol["x"]["0::" + stride];  // nk element starts + final state
  log("chariot position x at element starts (every 10th):");
  log("  " + xstarts["0::10"]);
  log("final x = " + xstarts[nk] + " (target xref = " + xref + ")");
}

if (typeof require !== "undefined" && typeof module !== "undefined" && require.main === module) {
  const path = require("path");
  const casadiPath = process.env.CASADI_JS
    || path.resolve(__dirname, "../../../build-wasm/swig/wasm-js/casadi.js");
  require(casadiPath)()
    .then((ca) => example(ca, (...a) => console.log(...a)))
    .catch((e) => { console.error("FATAL:", e.message || e); process.exit(1); });
}

if (typeof module !== "undefined" && module.exports) module.exports = example;

File: docs/examples/julia/dae_collocation.jl
#
#     MIT No Attribution
#
#     Copyright (C) 2010-2023 Joel Andersson, Joris Gillis, Moritz Diehl, KU Leuven.
#
#     Permission is hereby granted, free of charge, to any person obtaining a copy of this
#     software and associated documentation files (the "Software"), to deal in the Software
#     without restriction, including without limitation the rights to use, copy, modify,
#     merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#     permit persons to whom the Software is furnished to do so.
#
#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#     INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#     PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
# Hand-built direct collocation of a crane/pendulum DAE optimal control problem
# (Mario Zanon & Sebastien Gross, KU Leuven 2012), solved as one big NLP with
# ipopt. No integrator plugin is used -- the collocation equations are assembled
# by hand -- so this ports directly. The original ipopt option
# linear_solver='ma27' is dropped (default mumps is used).
const _jl_dir = get(ENV, "CASADI_JL",
    joinpath(@__DIR__, "..", "..", "..", "julia-proto", "casadi-contact"))
include(joinpath(_jl_dir, "CasADiNative.jl"))
import .CasADiNative as ca

function main()
    # -------- Collocation setup --------
    nicp = 1
    xref = 0.1
    l = 1.0; mmass = 1.0; Mmass = 1.0; g = 9.81
    tf = 5.0
    nk = 50
    ndstate = 6; nastate = 1; ninput = 1
    deg = 4
    h = tf / nk / nicp

    tau = ca.SX.sym("tau")
    tau_root = [0.0; ca.collocation_points(deg, "radau")]

    # Lagrange-polynomial collocation coefficients
    Cc = zeros(deg + 1, deg + 1)
    D = zeros(deg + 1)
    for j in 1:(deg + 1)
        L = ca.SX(ca.DM(1.0))
        for j2 in 1:(deg + 1)
            if j2 != j
                L = L * (tau - tau_root[j2]) / (tau_root[j] - tau_root[j2])
            end
        end
        lfcn = ca.Function("lfcn", [tau], [L])
        D[j] = Float64(lfcn(1.0))
        tfcn = ca.Function("tfcn", [tau], [ca.tangent(L, tau)])
        for j2 in 1:(deg + 1)
            Cc[j, j2] = Float64(tfcn(tau_root[j2]))
        end
    end

    # -------- Model setup (implicit DAE) --------
    t = ca.SX.sym("t")
    u = ca.SX.sym("u")
    xd = ca.SX.sym("xd", ndstate)
    xa = ca.SX.sym("xa", nastate)
    xddot = ca.SX.sym("xdot", ndstate)
    p = ca.SX.sym("p", 0, 1)

    x, y, w, dx, dy, dw = (xd[i] for i in 1:6)
    xa0 = xa[1]

    res = ca.vertcat(
        xddot[1] - dx,
        xddot[2] - dy,
        xddot[3] - dw,
        mmass * xddot[4] + (x - w) * xa0,
        mmass * xddot[5] + y * xa0 - g * mmass,
        Mmass * xddot[6] + (w - x) * xa0 + u,
        (x - w) * (xddot[4] - xddot[6]) + y * xddot[5] + dy * dy + (dx - dw) * (dx - dw))

    ffcn = ca.Function("ffcn", [t, xddot, xd, xa, u, p], [res])
    MayerTerm = ca.Function("mayer", [t, xd, xa, u, p],
        [(x - xref)^2 + (w - xref)^2 + dx * dx + dy * dy])
    LagrangeTerm = ca.Function("lagrange", [t, xd, xa, u, p],
        [(x - xref)^2 + (w - xref)^2])

    # Bounds
    u_min = [-2.0]; u_max = [2.0]
    xD_min = fill(-Inf, 6); xD_max = fill(Inf, 6)
    xDi_min = [0.0, l, 0.0, 0.0, 0.0, 0.0]; xDi_max = copy(xDi_min)
    xD_init = [0.0, l, 0.0, 0.0, 0.0, 0.0]
    xA_min = [-Inf]; xA_max = [Inf]; xA_init = [sign(l) * 9.81]

    # -------- NLP variable layout --------
    nx = ndstate + nastate
    ndiff = ndstate; nalg = nastate; nu = ninput; NP = 0
    NXD = nicp * nk * (deg + 1) * ndiff
    NXA = nicp * nk * deg * nalg
    NU = nk * nu
    NXF = ndiff
    NV = NXD + NXA + NU + NXF + NP

    V = ca.MX.sym("V", NV)
    vars_lb = zeros(NV); vars_ub = zeros(NV); vars_init = zeros(NV)
    setrange!(arr, off, vals) = (arr[(off + 1):(off + length(vals))] = vals)

    # XD[k][i][j], XA[k][i][j-1] (1-based), U[k] (slices of V)
    XD = [[Vector{Any}(undef, deg + 1) for _ in 1:nicp] for _ in 1:(nk + 1)]
    XA = [[Vector{Any}(undef, deg) for _ in 1:nicp] for _ in 1:nk]
    U = Vector{Any}(undef, nk)

    offset = 0
    for k in 1:nk
        for i in 1:nicp
            for j in 1:(deg + 1)
                XD[k][i][j] = V[(offset + 1):(offset + ndiff)]
                if j != 1
                    XA[k][i][j - 1] = V[(offset + ndiff + 1):(offset + ndiff + nalg)]
                end
                if k == 1 && j == 1 && i == 1
                    setrange!(vars_init, offset, xD_init)
                    setrange!(vars_lb, offset, xDi_min)
                    setrange!(vars_ub, offset, xDi_max)
                    offset += ndiff
                elseif j != 1
                    setrange!(vars_init, offset, vcat(xD_init, xA_init))
                    setrange!(vars_lb, offset, vcat(xD_min, xA_min))
                    setrange!(vars_ub, offset, vcat(xD_max, xA_max))
                    offset += nx
                else
                    setrange!(vars_init, offset, xD_init)
                    setrange!(vars_lb, offset, xD_min)
                    setrange!(vars_ub, offset, xD_max)
                    offset += ndiff
                end
            end
            U[k] = V[(offset + 1):(offset + nu)]
            setrange!(vars_lb, offset, u_min)
            setrange!(vars_ub, offset, u_max)
            setrange!(vars_init, offset, [0.0])
            offset += nu
        end
    end
    XD[nk + 1][1][1] = V[(offset + 1):(offset + ndiff)]
    setrange!(vars_lb, offset, xD_min)
    setrange!(vars_ub, offset, xD_max)
    setrange!(vars_init, offset, xD_init)
    offset += ndiff
    @assert offset == NV

    P = ca.MX(0, 1)

    # -------- Constraints: collocation + continuity --------
    g_con = ca.MX[]
    for k in 1:nk
        for i in 1:nicp
            for j in 2:(deg + 1)
                xp_jk = ca.MX(ca.DM(zeros(ndiff)))
                for j2 in 1:(deg + 1)
                    xp_jk = xp_jk + Cc[j2, j] * XD[k][i][j2]
                end
                fk = ffcn(ca.DM(0.0), xp_jk / h, XD[k][i][j], XA[k][i][j - 1], U[k], P)
                push!(g_con, fk[1:ndiff])               # differential part == 0
                push!(g_con, fk[(ndiff + 1):(ndiff + nalg)])  # algebraic part == 0
            end
            xf_k = ca.MX(ca.DM(zeros(ndiff)))
            for j in 1:(deg + 1)
                xf_k = xf_k + D[j] * XD[k][i][j]
            end
            if i == nicp
                push!(g_con, XD[k + 1][1][1] - xf_k)
            else
                push!(g_con, XD[k][i + 1][1] - xf_k)
            end
        end
    end

    # -------- Objective --------
    # Mayer term at the final collocation node
    Obj = MayerTerm(ca.DM(0.0), XD[nk][1][deg + 1], XA[nk][1][deg], U[nk], P)

    # Lagrange term via the collocation quadrature weights:
    #   lDotAtTauRoot = ca.T ; ldInv = inv(ca.T[2:,2:]) ; lAtOne = D[2:]
    CT = collect(transpose(Cc))
    ldInv = ca.MX(ca.inv(ca.DM(CT[2:end, 2:end])))
    lAtOne1 = ca.MX(ca.DM(D[2:end]))
    for k in 1:nk
        for i in 1:nicp
            dqParts = ca.MX[]
            for j in 2:(deg + 1)
                push!(dqParts, LagrangeTerm(ca.DM(0.0), XD[k][i][j], XA[k][i][j - 1], U[k], P))
            end
            dQs = h * ca.vertcat(dqParts)               # deg x 1
            Qs = ca.mtimes(ldInv, dQs)                # deg x 1
            Obj = Obj + ca.mtimes(Qs', lAtOne1)  # scalar
        end
    end

    # -------- Solve --------
    nlp = Dict("x" => V, "f" => Obj, "g" => ca.vertcat(g_con))
    solver = ca.nlpsol("solver", "ipopt", nlp, Dict(
                    "expand" => true, "ipopt.tol" => 1e-4,
                    "ipopt.print_level" => 0, "print_time" => false))
    sol = solver(x0 = vars_init, lbx = vars_lb, ubx = vars_ub, lbg = 0, ubg = 0)

    println("optimal cost: ", Float64(sol["f"]))

    # Chariot position x (xd[1]) at the start of each finite element: fixed stride
    stride = nicp * ((deg + 1) * ndiff + deg * nalg) + nu
    xstarts = sol["x"][1:stride:end]   # nk element starts + final state
    println("chariot position x at element starts (every 10th):")
    println("  ", xstarts[1:10:end])
    println("final x = ", xstarts[nk + 1], " (target xref = ", xref, ")")
end

main()
